sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java

```diff
@@ -854,35 +858,43 @@ private CompletableFuture<PrepareResult> prepareResourceAsync(
+
+                                                                // Construct the provider reference, if we were given a provider to use.
+                                                                final CompletableFuture<Optional<String>> providerRef;
+                                                                if (custom) {
+                                                                    providerRef = CompletableFutures.flipOptional(options.getProvider().map(p -> Internal.from(p).getRegistrationId()));
+                                                                } else {
+                                                                    providerRef = CompletableFuture.completedFuture(Optional.<String>empty());
+                                                                }
 
                                                                 return providerRef.thenCompose(
                                                                         (Optional<String> pRef) -> {
+                                                                            // For remote resources, merge any provider opts into a single dict,
+                                                                            // and then create a new dict with all the resolved provider refs.
                                                                             final CompletableFuture<ImmutableMap<String, String>> providerFutures;
                                                                             if (remote && options instanceof ComponentResourceOptions) {
                                                                                 var componentOpts = (ComponentResourceOptions) options;
 
-                                                                                // TODO: C# had the following logic here:
-                                                                                //          "If only the Provider opt is set, move it to the Providers list for further processing."
-                                                                                //       But our ComponentResourceOptions should guarantee the desired semantics.
-                                                                                //       It would be great to add more tests and maybe remove 'provider' in favour of 'providers' only
+                                                                                log.excessive("Processing a remote ComponentResource: t=%s, name=%s, custom=%s, remote=%s", type, name, custom, remote);
+
+                                                                                BiFunction<Optional<ProviderResource>, List<ProviderResource>, Map<String, CompletableFuture<String>>> convertProviders =
+                                                                                        (provider, providers) -> Stream.concat(provider.stream(), providers.stream())
```

Comment:
Prepending `provider.stream()` to `providers.stream()` is the only behavioral change in this PR, but it should not matter since we handle this in Resource constructor if I'm not mistaken.

I've decided that adding this for completeness is probably not a bad idea.

Reply:
Unfortunately it's possibly a semantic change - I don't understand this change therefore cannot approve this; we'd need tests to clarify what's going on, and possibly reference to another SDK to show the desired behavior this change is trying to achieve.

What would be really helpful is to detangle this. It seems that:

- we are computing RegisterResourceRequest Providers field (this is our output) 
- we have as an input ComponentResourceOptions.getProviders
- there is tricky interaction with `provider` field, so possibly that is an input as well
- not clear what's responsibility of this code vs resource constructor - can you clarify that comment? 
-  ealing with this at 12 level of indent is not helpful

Let me ask @iwahbe about related work on provider vs provider and where we landed with that. I found https://docs.google.com/document/d/1Z0P-XU5AW-dhwKQCWHEiau_7kIAsF9j4CUFaPlMlGj4/edit#heading=h.4brqwrn4u47v but I need to check what the state of the project is.

Reply:
I've filed a new issue that represents better how Ian's work actually ended unifying provider/providers across SDKs. There's some good stuff here. I suggest we treat it as a separate work item and do thoroughly when time permits. Tyvm. The issue is #681